File: apps/trak/models/handler.py
from django.db import models


class Handler(models.Model):
    epa_id = models.CharField(
        verbose_name='EPA Id number',
        max_length=25,
    )
    name = models.CharField(
        max_length=200,
    )
    site_address = models.ForeignKey(
        'Address',
        on_delete=models.CASCADE,
        related_name='site_address',
    )
    mail_address = models.ForeignKey(
        'Address',
        on_delete=models.CASCADE,
        related_name='mail_address',
    )
    modified = models.BooleanField(
        null=True,
        blank=True,
    )
    registered = models.BooleanField(
        null=True,
        blank=True,
    )
    contact = models.JSONField(
        verbose_name='Contact information')
    emergency_phone = models.JSONField(
        null=True,
        blank=True,
    )
    electronic_signatures_info = models.JSONField(
        verbose_name='Electronic signature info',
        null=True,
        blank=True,
    )
    gis_primary = models.BooleanField(
        verbose_name='GIS primary',
        null=True,
        blank=True,
        default=False,
    )
    can_esign = models.BooleanField(
        verbose_name='Can electronically sign',
        null=True,
        blank=True,
    )
    limited_esign = models.BooleanField(
        verbose_name='Limited electronic signing ability',
        null=True,
        blank=True,
    )
    registered_emanifest_user = models.BooleanField(
        verbose_name='Has Registered e-Manifest user',
        null=True,
        blank=True,
        default=False,
    )

    def __str__(self):
        return f'{self.epa_id}'


class Site(models.Model):
    name = models.CharField(
        verbose_name='Site Alias',
        max_length=200,
    )
    epa_site = models.OneToOneField(
        verbose_name='Handler',
        to=Handler,
        on_delete=models.CASCADE,
    )

    def get_absolute_url(self):
        from django.urls import reverse
        return reverse('site_details', kwargs={'pk': self.pk})

    def __str__(self):
        return f'{self.epa_site.epa_id}'

File: client/src/types/handler/handler.ts
import { Address, Locality } from './address';
import { Contact, Phone } from 'types/handler';

/**
 * The schema of an EPA hazardous waste handler, it serves as the base schema for
 * other schemas
 */
export interface Handler {
  name: string;
  /**
   * The RCRAInfo administered EPA ID used by registered Generators, Transporters and
   * designated receiving facilities (AKA "TSDF") listed on a hazardous waste manifest.
   * Follows a predictable format: Two Letter Activity Location Code + Up to 10 alphanumeric characters
   */
  epaSiteId: string;
  mailingAddress: Address;
  siteAddress: Address;
  contact: Contact;
  /**
   * The phone number to call if an emergency occurs during hazardous waste shipment
   */
  emergencyPhone: Phone;
  /**
   * Whether the handler has a registered user in RCRAInfo
   */
  registered?: boolean;
  modified?: boolean;
  /**
   * Indicates there a registered user with 'preparer' level access associated with the handler
   * who can 'quick sign'.
   */
  limitedEsign?: boolean;
  /**
   * Indicates there is a user associated with the site which can electronically sign
   */
  canEsign?: boolean;
  /**
   * The U.S. geographic state of the handler
   */
  siteState?: Locality;
  /**
   *
   * Indicates there is a registered user associated with the site. A registered user does
   * not necessarily mean a registered user who can electronically sign.
   */
  hasRegisteredEmanifestUser?: boolean;
  /**
   * Indicates if a site's address1 uses latitude/longitude
   */
  gisPrimary?: boolean;
}

/**
 * The Signature that appears on paper versions of the manifest
 */
export interface PaperSignature {
  printedName: string;
  signatureDate: string;
}

/**
 * The ManifestHandler extends to Handler schema and adds manifest specific data
 */
export interface ManifestHandler extends Handler {
  paperSignatureInfo?: PaperSignature;
  electronicSignaturesInfo?: Array<ElectronicSignature>;
  /**
   * Property on by back end to signify whether the handler has signed
   */
  signed?: boolean;
}

/**
 *  The Transporter type extends the ManifestHandler schema
 */
export interface Transporter extends ManifestHandler {
  order: number;
  manifest?: number;
}

/**
 *  The Site type encapsulates the Handler model as well as other details specific to
 *  that hazardous waste site. It is not directly associated with the EPA RCRAInfo system
 *  directly and can be extended as needed. For example, Haztrak users are given permission to
 *  sites, not Handlers.
 */
export interface Site {
  name: string;
  handler: Handler;
}

/**
 * EPA schema of an electronic signature
 */
export interface ElectronicSignature {
  signer?: Signer;
  signatureDate?: string;
  /**
   * Object representing metadata on a printable, human friendly, version of the manifest
   * ToDo: implement this schema, see the USEPA/e-Manifest documentation on GitHub
   */
  humanReadableDocument?: object;
}

/**
 * Used to specify whether a handler is a generator, transporter, or
 * designated receiving facility (AKA Treatment, Storage and Disposal Facility or TSD/TSDF for short).
 */
export enum HandlerType {
  Generator = 'generator',
  Tsd = 'designatedFacility',
  Transporter = 'transporter',
}

/**
 * EPA definition of a manifest signer
 */
export interface Signer {
  /**
   * User's RCRAInfo username
   */
  userId?: string;
  firstName?: string;
  middleInitial?: string;
  lastName?: string;
  phone?: Phone;
  email?: string;
  companyName?: string;
  contactType?: 'Email' | 'Text' | 'Voice';
}
